CBS script enhancements to manage packages lifecycle and ability to fetch external dependencies (#299)
IBM/dbb: Templates/Common-Backend-Scripts/README.md

## Setup

The provided scripts of this asset are implemented as bash scripts. The entire repository needs to be installed on UNIX System Services of the z/OS system that is used to execute the pipeline's tasks, because as previously mentioned the scripts invoke individual pipeline scripts from the [Pipeline](../../Pipeline/) directory.

### Pre-requisites
The following are required to use these scripts:
* DBB v2.x / 3.x toolkit is installed.
* zAppBuild or zBuilder is set up on Unix Systems Services.
* Git repository which follows the Git-based workflow outlined in IBM's documentation `The Git-based workflow for Mainframe development`.

### Installation

> * Make this Git repository available in your internal git provider. Perform necessary site-specific [script configuration](#script-configuration).

Reviewer: Should we provide some brief guidance on how the reader can make the Git repo available in their enterprise Git provider? For example:
> 1. Make a customized copy of this Git repository available in your enterprise Git provider:
>    1. Follow your company policy to create an empty repository in your enterprise Git server.
>    2. Copy the contents of this repository ([the DBB community repository](https://github.com/IBM/dbb)) into the empty repository from the previous step.
>    3. Update your copy of the repository with necessary site-specific [script configurations](#script-configuration).

I believe this bullet point and the following one ("Clone the repository...") should also be numbered, since they need to be performed in a specific order.

> * Make this Git repository available in your internal git provider. Perform necessary site-specific [script configuration](#script-configuration).

Author: We want to take a fork of this repo rather than copying the files (leading to potential code page issues).

> * Make this Git repository available in your internal git provider. Perform necessary site-specific [script configuration](#script-configuration).

Reviewer: @dennis-behm Thank you for the clarification! I don't believe it's quite straightforward (especially for newcomers to Git) to fork the repo from the public GitHub to an enterprise Git server. Should we maybe link to some advice on how to do this? (Maybe something like our [install/config page](https://ibm.github.io/z-devops-acceleration-program/docs/install-configure#making-zappbuild-available-in-your-git-provider)... except apply the instructions to CBS instead of zAppBuild?)

> * Make this Git repository available in your internal git provider. Perform necessary site-specific [script configuration](#script-configuration).

Author: @lauren-li Thanks for pointing that out. I would like to treat the enhancement of the install steps as a separate issue and don't overload this PR. Do you agree?

> * Make this Git repository available in your internal git provider. Perform necessary site-specific [script configuration](#script-configuration).

Reviewer: @dennis-behm Yes, I think it's better to enhance the install steps in a future PR, as it's quite separate from the main focus of this PR.

> * Clone the repository that include the Common Backend Scripts into z/OS UNIX System Services under a protected directory.

Reviewer: ```suggestion
2. Clone the repository customized in the previous step (including the Common Backend Scripts) into z/OS UNIX System Services under a protected directory.
```

[baselineReference.config](samples/baselineReference.config) is a sample, that indicates the baseline for the `main` and `release maintenance` branches.

#### Fetching build dependencies

Both implementations for the build stage are enabled to optionally pull external dependencies into the build workspace based on the dependency definitions specified in the *Application Descriptor*.

> The Application Descriptor contains metadata about the application itself, but can contain references to other application packages managed in an artifact repository that are necessary inputs to the build. Please read more about the Application Descriptor at the [dbb-git-migration-modeler](https://github.com/IBM/dbb-git-migration-modeler) project, which generates it and provides insights about cross application dependencies.

Reviewer 2: ```suggestion
The Application Descriptor contains metadata about the application itself, but can contain references to other application packages managed in an artifact repository, which are necessary inputs to the build. Additional information about the Application Descriptor can be found at the [dbb-git-migration-modeler](https://github.com/IBM/dbb-git-migration-modeler) project, which documents cross-application dependencies and generates Application Descriptor files.
```

> Each application package can export shared components such as shared copybooks and even build outputs such as object decks or NCAL load modules. The package needs to be created with the [PackageBuildOutputs](../../Pipeline/PackageBuildOutputs/README.md) script and be uploaded to the Artifact repository through the Common Backend scripts. Fetching the dependencies is powered on the [ArtifactoryHelpers](../../Pipeline/PackageBuildOutputs/ArtifactRepositoryHelpers.groovy)

Reviewer 2: ```suggestion
Each application package can export shared components such as public or shared include files and even build outputs, such as object decks or NCAL load modules. The package needs to be created with the [PackageBuildOutputs](../../Pipeline/PackageBuildOutputs/README.md) script and be uploaded to the artifact repository through the Common Backend scripts. Fetching the dependencies is powered by the [ArtifactoryHelpers](../../Pipeline/PackageBuildOutputs/ArtifactRepositoryHelpers.groovy) script.
```
